tests/test_properties.py

```diff
@@ -0,0 +1,109 @@
+
+from zarr._storage.v3 import KVStoreV3
+from zarr.storage import init_array, normalize_store_arg
+
+#### TODO: Provide this in zarr.strategies
+# Copied from Xarray
+# only use characters within the "Latin Extended-A" subset of unicode
+_readable_characters = st.characters(categories=["L", "N"], max_codepoint=0x017F)
+_readable_strings = st.text(_readable_characters, max_size=5)
+_attr_keys = st.text(_readable_characters, min_size=1)
+_attr_values = st.recursive(
+    st.none() | st.booleans() | _readable_strings,
+    lambda children: st.lists(children) | st.dictionaries(_attr_keys, children),
+    max_leaves=3,
+)
+
+# No '/' in array names?
+# No '.' in paths?
+zarr_key_chars = st.sampled_from("-0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz")
+
+# The following should be public strategies
+attrs = st.none() | st.dictionaries(_attr_keys, _attr_values)
+paths = st.none() | st.text(zarr_key_chars, min_size=1) | st.just("/")
+array_names = st.text(zarr_key_chars | st.just("."), min_size=1).filter(
+    lambda t: not t.startswith((".", ".."))
+)
+
+
+@st.composite
+def array_strategy(draw):
+    """A hypothesis strategy to generate small sized random arrays.
+
+    Returns: a tuple of the array and a suitable random chunking for it.
+    """
+    array = draw(npst.arrays(dtype=npst.scalar_dtypes(), shape=npst.array_shapes(max_dims=4)))
+    # We want this strategy to shrink towards arrays with smaller number of chunks
+    # 1. st.integers() shrinks towards smaller values. So we use that to generate number of chunks
+    numchunks = draw(st.tuples(*[st.integers(min_value=1, max_value=size) for size in array.shape]))
+    # 2. and now generate the chunks tuple
+    chunks = tuple(size // nchunks for size, nchunks in zip(array.shape, numchunks))
+    return (array, chunks)
+
+
+#####
+
+
+# @pytest.mark.slow
+@settings(max_examples=300)
+@given(path=paths, attrs=attrs, name=array_names, array_and_chunks=array_strategy())
+def test_roundtrip(path, array_and_chunks, name, attrs):
+    store = KVStoreV3(dict())
+
+    nparray, chunks = array_and_chunks
+
+    # TODO: clean this up
+    if path is None and name is None:
+        array_path = None
+        array_name = None
+    elif path is None and name is not None:
+        array_path = f"{name}"
+        array_name = f"/{name}"
+    elif path is not None and name is None:
+        array_path = path
+        array_name = None
+    elif path == "/":
+        assert name is not None
+        array_path = name
+        array_name = "/" + name
+    else:
+        assert name is not None
+        array_path = f"{path}/{name}"
+        array_name = "/" + array_path
+
+    expected_attrs = {} if attrs is None else attrs
+
+    init_array(store, shape=nparray.shape, chunks=chunks, path=array_path, dtype=nparray.dtype.str)
+    a = Array(store, path=array_path, zarr_version=3)
+    if attrs is not None:
+        a.attrs.put(attrs)
+
+    assert isinstance(a, Array)
```

Comment:
I'd probably move all of this into a `zarr_arrays()` function, which returns an arbitrary Zarr array.

The general design pattern is that the strategy should generate arbitrarily weird and nasty instances by default, and then you can provide various arguments to constrain it (e.g., the arguments you've passed to `@given()` here).  For example, why constrain names to Latin Extended-A?  If that's part of the Zarr spec it'd be nice to have a comment, if not we should check for bugs in the rest of utf-8 at least (if not all of Unicode).

Reply:
Ah yes, I agree.

I'm a bit confused with how to do that since Arrays need to be backed by stores, but I can think about it.

Good point about the unicode thing, that was an Xarray restriction that is not needed here.

Reply:
`zarr_arrays(..., store: st.SearchStrategy[store] = ...)`?  You can pass a specific store with `st.just()`, and access a generated store via the `arr.store` attribute.